File: 0725-split-linked-list-in-parts/0725-split-linked-list-in-parts.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int length(ListNode* head) {
        int cnt = 0;
        while (head != nullptr) {
            ++cnt;
            head = head->next;
        }
        return cnt;
    }

    vector<ListNode*> splitListToParts(ListNode* head, int k) {
        vector<ListNode*> res(k);
        const int list_len = length(head);
        const int short_cnt = list_len % k == 0 ? list_len % k : k - list_len % k;
        const int long_num = (list_len + (k - 1)) / k;
        for (int i = 0; i < k - short_cnt; ++i) {
            res[i] = head;
            for (int j = 0; j < long_num - 1; ++j) {
                head = head->next;
            }
            if (long_num > 0) {
                ListNode* prev = head;
                head = head->next;
                prev->next = nullptr;
            }
        }
        for (int i = 0; i < short_cnt; ++i) {
            res[k - short_cnt + i] = head;
            for (int j = 0; j < long_num - 2; ++j) {
                head = head->next;
            }
            if (long_num - 1 > 0) {
                ListNode* prev = head;
                head = head->next;
                prev->next = nullptr;
            }
        }

        return res;
    }
};

File: 0725-split-linked-list-in-parts/0725-split-linked-list-in-parts.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
use std::mem;

impl Solution {
    pub fn length_of_list(mut head: &Option<Box<ListNode>>) -> usize {
        let mut len = 0;
        while let Some(node) = head {
            len += 1;
            head = &node.next;
        }
        len
    }

    pub fn split_list_to_parts(mut head: Option<Box<ListNode>>, k: i32) -> Vec<Option<Box<ListNode>>> {
        let list_length = Solution::length_of_list(&head);
        let k = k as usize;
        let long_list_len = if list_length == 0 {
            0
        } else {
            1 + (list_length - 1) / k
        };

        let mut long_list_cnt = list_length % k;
        if long_list_cnt == 0 {
            long_list_cnt = k;
        }
        let mut res = vec![None; k];
        if long_list_len > 0 {
            for i in 0..long_list_cnt {
                res[i] = Some(Box::new(ListNode::new(head.as_ref().unwrap().val)));
                let mut last_elem = res[i].as_mut().unwrap();
                for j in 0..long_list_len - 1 {
                    let node = head.unwrap();
                    head = node.next;
                    last_elem.next = Some(Box::new(ListNode::new(head.as_ref().unwrap().val)));
                    last_elem = last_elem.next.as_mut().unwrap();
                }

                let mut node = head.unwrap();
                head = node.next;
                node.next = None;
            }
        }
        
        if long_list_len - 1 > 0 {
            for i in long_list_cnt..k {
                res[i] = Some(Box::new(ListNode::new(head.as_ref().unwrap().val)));
                let mut last_elem = res[i].as_mut().unwrap();
                for j in 0..long_list_len - 2 {
                    let node = head.unwrap();
                    head = node.next;
                    last_elem.next = Some(Box::new(ListNode::new(head.as_ref().unwrap().val)));
                    last_elem = last_elem.next.as_mut().unwrap();
                }

                let mut node = head.unwrap();
                head = node.next;
                node.next = None;
            }
        }

        res
    }
}